### PR TITLE
Restore baseFee opcode

### PR DESCRIPTION
**Pull request by gastonponti**
### Description

Implementation of the CIP-61: Restore BaseFee opcode/header field.

### Tested

Run test, e2e tests, and spin a testnet checking the hf switch in a heavy load

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

This is part of the Gingerbread hardfork

### CODE REVIEW

File: mycelo/genesis/genesis.go
Comment (karlb) [inline mycelo/genesis/genesis.go:24]: Is this something we really want, or is it just included temporarily for testing the change?

File: accounts/abi/bind/backends/simulated.go
Comment (karlb) [inline accounts/abi/bind/backends/simulated.go:536]: Why is this necessary to add the baseFee header field or opcode?

File: eth/gasprice/gasprice.go
Comment (karlb) [inline eth/gasprice/gasprice.go:35]: Not sure what the most relevant convention is, but I'm used reading it as either
```go
	if baseFee != nil {  // Gingerbread Fork
```
or
```go
	if baseFee != nil {
        // Gingerbread Fork
        ...
```

File: eth/gasprice/gasprice.go
Comment (karlb) [inline eth/gasprice/gasprice.go:42]: Is zero a good fallback here? Shouldn't we make transactions expensive rather than free if in doubt?

File: eth/gasprice/gasprice.go
Comment (karlb) [inline eth/gasprice/gasprice.go:48]: We could reduce the nesting of most of the function if we started the function with
```go
if baseFee == nil {
	return gpm.GetGasPriceMinimum(vmRunner, currencyAddress)
}
```
I also prefer to have the code for old forks in the `if` and the current fork on lowest indentation.

**Review by karlb — COMMENTED** [review]
This got bigger than I originally anticipated. Looks good so far. Feel free to ignore the nitpicks.

File: eth/gasprice/gasprice.go
Comment (gastonponti) [reply eth/gasprice/gasprice.go:35]: I this case, as I'm going to make the other recommendation of the `if baseFee == nil {`, I'll keep it after the if

File: e2e_test/e2e_transfer_test.go
Comment (gastonponti) [reply e2e_test/e2e_transfer_test.go:34]: No, it was used a way of debugging, but I'll remove it

File: mycelo/genesis/genesis.go
Comment (gastonponti) [reply mycelo/genesis/genesis.go:24]: Temporary for testing

**Comment by palango** [comment]
What's the relation of this PR and https://github.com/celo-org/celo-monorepo/pull/10274? Do we need to wait until the update is merged or do we just pull the changes in our monorepo branch? If so, will that cause problems with the v10 contracts release?

File: core/vm/jump_table.go
Comment (palango) [inline core/vm/jump_table.go:70]: nit: From the name and ordering I'd expect that below `enable3529`.

File: accounts/abi/bind/backends/simulated.go
Comment (gastonponti) [reply accounts/abi/bind/backends/simulated.go:536]: This is something I've bringed from upstream to be as close as possible, but is something that I can remove if we are not confident that we want it

Comment (karlb) [reply accounts/abi/bind/backends/simulated.go:536]: It is fine with me, I was just irritated because I didn't see the reason for the change. That's clearer now.

File: eth/gasprice/gasprice.go
Comment (gastonponti) [reply eth/gasprice/gasprice.go]: Yes, haha, agree on that. I've added it here as zero because it was the same behaviour we had on production. Technically speaking, if we don't have the gasPriceMinimum contract nor the sorted oracles of a stable, we are probably in a bigger problem, but yes, we could change the default on how we handle this

Comment (karlb) [reply eth/gasprice/gasprice.go]: I agree that his is a rare edge case and the specific handling is not too important. We can leave it as zero to minimize the changes for now, if you think that is safer.

File: consensus/misc/eip1559.go
Comment (gastonponti) [reply consensus/misc/eip1559.go:54]: Yes, merged from master the gasLimit change in the header and keep the old one for testing. I'll remove it

File: core/state_processor_test.go
Comment (gastonponti) [reply core/state_processor_test.go:334]: This badBlock function stays that "the block is sufficiently valid to be considered for import".
Everything there is mocked. In our scenario, as we require a runner to calculate the gasPrice, would have being changing a lot for a field that is not important in the tests that are invoking it.
Also I've noticed that we have the creation of the Extra field, twice (I'll check if it works without the second set, which should, and remove that line)

**Review by karlb — APPROVED** [review]

File: core/types/block.go
Comment (mcortesi) [inline core/types/block.go:357]: what if h.BaseFee is nil?

File: cmd/mycelo/genesis_commands.go
Comment (mcortesi) [inline cmd/mycelo/genesis_commands.go:102]: is this ok?

Comment (gastonponti) [reply cmd/mycelo/genesis_commands.go:102]: Yes, the monorepo changed their build folder to have `contracts`, and `contracts-mento`. As they have some contracts that collide with their names, I preferred to change our mycelo implementation to decide which folder to use

File: core/types/block.go
Comment (gastonponti) [reply core/types/block.go:357]: The cpy will still have a nil as baseFee. I don't know if I understand the question

**Pull request merged**
